docs/proposals/protobuf.md

```diff
@@ -0,0 +1,501 @@
+    JSON             80allocs               100allocs
+    Proto                                   4allocs
+
+ Protobuf has a huge benefit on encoding because it does not need to allocate
+ temporary objects, just one large buffer. Changing to protobuf moves our
+ hotspot back to conversion, not serialization.
+
+
+## Design Points
+
+* Generate Protobuf schema from Go structs (like we do for JSON) to avoid
+  manual schema update and drift
+* Generate Protobuf schema that is field equivalent to the JSON fields (no
+  special types or enumerations), reducing drift for clients across formats.
+* Follow our existing API versioning rules (backwards compatible in major
+  API versions, breaking changes across major versions) by creating one
+  Protobuf schema per API type.
+* Continue to use the existing REST API patterns but offer an alternative
+  serialization, which means existing client and server tooling can remain
+  the same while benefiting from faster decoding.
+* Protobuf objects on disk or in etcd will need to be self identifying at
+  rest, like JSON, in order for backwards compatibility in storage to work,
+  so we must add an envelope with apiVersion and kind to wrap the nested
+  object, and make the data format recognizable to clients.
+* Use the [gogo-protobuf](https://github.com/gogo/protobuf) Golang library to generate marshal/unmarshal
+  operations, allowing us to bypass the expensive reflection used by the
+  golang JSOn operation
+
+
+## Alternatives
+
+* We considered JSON compression to reduce size on wire, but that does not
+  reduce the amount of memory garbage created during serialization and
+  deserialization.
+* More efficient formats like Msgpack were considered, but they only offer
+  2x speed up vs the 10x observed for Protobuf
+* gRPC was considered, but is a larger change that requires more core
+  refactoring. This approach does not eliminate the possibility of switching
+  to gRPC in the future.
+* We considered attempting to improve JSON serialization, but the cost of
+  implementing a more efficient serializer library than ugorji is
+  significantly higher than creating a protobuf schema from our Go structs.
+
+
+## Schema
+
+The Protobuf schema for each API group and version will be generated from
+the objects in that API group and version. The schema will be named using
+the package identifier of the Go package, i.e.
+
+    k8s.io/kubernetes/pkg/api/v1
+
+Each top level object will be generated as a Protobuf message, i.e.:
+
+    type Pod struct { ... }
+
+    message Pod {}
+
+Since the Go structs are designed to be serialized to JSON (with only the
+int, string, bool, map, and array primitive types), we will use the
+canonical JSON serialization as the protobuf field type wherever possible,
+i.e.:
+
+    JSON      Protobuf
+    string -> string
+    int    -> varint
+    bool   -> bool
+    array  -> repeating message|primitive
+
+We disallow the use of the Go `int` type in external fields because it is
```

Comment:
The checker would be that protobuf generation would fail (because go2idl
won't be able to pick an unambiguous type).  I'll need to double check that
we catch that earlier.

On Mon, Mar 7, 2016 at 4:03 AM, Wojciech Tyczynski <notifications@github.com
> wrote:

> In docs/proposals/protobuf.md
> <https://github.com/kubernetes/kubernetes/pull/22600#discussion_r55182327>
> :
>
> > +    type Pod struct { ... }
> > +
> > +    message Pod {}
> > +
> > +Since the Go structs are designed to be serialized to JSON (with only the
> > +int, string, bool, map, and array primitive types), we will use the
> > +canonical JSON serialization as the protobuf field type wherever possible,
> > +i.e.:
> > +
> > +    JSON      Protobuf
> > +    string -> string
> > +    int    -> varint
> > +    bool   -> bool
> > +    array  -> repeating message|primitive
> > +
> > +We disallow the use of the Go `int` type in external fields because it is
>
> I remember that you were switching "int" to "int32" and "int64" at some
> point. But do we have some checker that is testing it e.g. on every commit?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/22600/files#r55182327>.
>